spec/googleauth/impersonated_service_account_spec.rb

```diff
@@ -0,0 +1,144 @@
+# Copyright 2024 Google, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+spec_dir = File.expand_path File.join(File.dirname(__FILE__))
+$LOAD_PATH.unshift spec_dir
+$LOAD_PATH.uniq!
+
+require "googleauth/impersonated_service_account"
+require "spec_helper"
+
+describe Google::Auth::ImpersonatedServiceAccountCredentials do
+
+  let(:impersonation_url) {"https://iamcredentials.example.com/v1/projects/-/serviceAccounts/test:generateAccessToken"}
+
+  def make_auth_stubs opts
+    body_fields = { "token_type" => "Bearer", "expires_in" => 3600 }
+    body_fields["accessToken"] = opts[:access_token]
+    body_fields["expireTime"] = opts[:expireTime]
+    body = MultiJson.dump body_fields
+    stub_request(:post, impersonation_url)
+      .to_return(body:    body,
+                   status:  200,
+                   headers: { "Content-Type" => "application/json" })
+  end
+
+  before :example do
+    @base_creds = double("Credentials")
+  end
+
+  describe "#initialize" do
+    it "should call duplicate when available" do
+      allow(@base_creds).to receive(:duplicate).and_return(@base_creds)
```

Comment:
maybe return something else so you can verify that `source_credentials` is that other thing rather than `@base_creds`.